File: TUI-ConsoleLauncher/app/src/main/java/ohi/andre/consolelauncher/commands/smartlauncher/productivity/NotesCommand.java
package ohi.andre.consolelauncher.commands.smartlauncher.productivity;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import ohi.andre.consolelauncher.commands.CommandAbstraction;
import ohi.andre.consolelauncher.commands.ExecutePack;

/**
 * Notes Command - Quick note taking and retrieval
 * Supports creating, listing, searching, and managing notes
 */
public class NotesCommand implements CommandAbstraction {

    private static final String TAG = "NotesCommand";
    private static final String PREFS_NAME = "tui_notes_prefs";
    private static final String KEY_NOTES = "saved_notes";
    private static final int MAX_NOTE_LENGTH = 10000;

    /**
     * Note data class equivalent
     */
    static class Note {
        String id;
        String title;
        String content;
        long createdAt;
        long updatedAt;
        ArrayList<String> tags;

        Note(String id, String title, String content, long createdAt, long updatedAt, ArrayList<String> tags) {
            this.id = id;
            this.title = title;
            this.content = content;
            this.createdAt = createdAt;
            this.updatedAt = updatedAt;
            this.tags = tags;
        }
    }

    @Override
    public String exec(ExecutePack pack) throws Exception {
        Context context = pack.context;
        
        ArrayList<String> argsList = new ArrayList<>();
        for (Object arg : pack.args) {
            if (arg != null) {
                argsList.add(arg.toString());
            }
        }

        if (argsList.isEmpty()) {
            return listNotes(context);
        }

        String command = argsList.get(0).toLowerCase();
        ArrayList<String> parameters = new ArrayList<>(argsList.subList(1, argsList.size()));

        switch (command) {
            case "create":
            case "new":
            case "add":
                return createNote(context, parameters);
            case "list":
            case "ls":
                return listNotes(context);
            case "show":
            case "view":
            case "get":
                return showNote(context, parameters);
            case "edit":
            case "update":
            case "modify":
                return editNote(context, parameters);
            case "delete":
            case "del":
            case "remove":
                return deleteNote(context, parameters);
            case "search":
            case "find":
            case "grep":
                return searchNotes(context, parameters);
            case "tag":
            case "tags":
                return tagNote(context, parameters);
            case "export":
                return exportNote(context, parameters);
            case "clear":
                return clearAllNotes(context);
            default:
                // Try to find note by title or ID
                return showNote(context, argsList);
        }
    }

    @Override
    public int[] argType() {
        return new int[]{CommandAbstraction.COMMAND};
    }

    @Override
    public int priority() {
        return 5;
    }

    @Override
    public int helpRes() {
        return 0;
    }

    @Override
    public String onArgNotFound(ExecutePack pack, int indexNotFound) {
        return "Missing argument at position " + indexNotFound + "\n" + getUsage();
    }

    @Override
    public String onNotArgEnough(ExecutePack pack, int nArgs) {
        return "Not enough arguments (" + nArgs + " required)\n" + getUsage();
    }

    private String getUsage() {
        return "\n╔══════════════════════════════════════════════════════╗\n" +
               "║                   NOTES COMMANDS                      ║\n" +
               "╠══════════════════════════════════════════════════════╣\n" +
               "║  note create <title>         - Create new note       ║\n" +
               "║  note create <title> <text>  - Create with content   ║\n" +
               "║  note list                   - List all notes        ║\n" +
               "║  note show <id>              - Show note content     ║\n" +
               "║  note show <title>           - Show by title         ║\n" +
               "║  note edit <id> <text>       - Edit note content     ║\n" +
               "║  note delete <id>            - Delete a note         ║\n" +
               "║  note search <query>         - Search notes          ║\n" +
               "║  note tag <id> <tag>         - Add tag to note       ║\n" +
               "║  note export <id>            - Export note as text   ║\n" +
               "╚══════════════════════════════════════════════════════╝\n";
    }

    /**
     * Creates a new note
     */
    private String createNote(Context context, ArrayList<String> args) {
        if (args.isEmpty()) {
            return "Usage: note create <title> [content]\nUse 'note create \"My Title\" This is content' for inline creation";
        }

        String title = args.get(0);
        String content = "";
        if (args.size() > 1) {
            StringBuilder contentBuilder = new StringBuilder();
            for (int i = 1; i < args.size(); i++) {
                if (i > 1) contentBuilder.append(" ");
                contentBuilder.append(args.get(i));
            }
            content = contentBuilder.toString();
        }

        try {
            long now = System.currentTimeMillis();
            Note note = new Note(
                generateNoteId(),
                title,
                content,
                now,
                now,
                extractTags(content)
            );

            saveNote(context, note);
            
            String preview = content.length() > 50 ? content.substring(0, 50) + "..." : content;
            String contentInfo = !content.trim().isEmpty() ? 
                "Content: " + preview : "(Empty note - use 'note edit " + note.id + "' to add content)";
            
            return "Note created: [" + note.id + "] " + note.title + "\n" + contentInfo;
        } catch (Exception e) {
            Log.e(TAG, "Error creating note", e);
            return "Error creating note: " + e.getMessage();
        }
    }

    /**
     * Lists all notes
     */
    private String listNotes(Context context) {
        ArrayList<Note> notes = getAllNotes(context);
        if (notes.isEmpty()) {
            return "\nNo notes yet.\n\nCreate your first note:\n  note create \"My Title\" This is the content";
        }

        StringBuilder builder = new StringBuilder();
        builder.append("\n");
        builder.append("═══════════════════════════════════════════════════════════════\n");
        builder.append("                         NOTES (").append(notes.size()).append(")                            \n");
        builder.append("═══════════════════════════════════════════════════════════════\n");

        // Sort by updated date (newest first)
        notes.sort((n1, n2) -> Long.compare(n2.updatedAt, n1.updatedAt));

        for (int index = 0; index < notes.size(); index++) {
            Note note = notes.get(index);
            String preview = note.content.replace("\n", " ");
            if (preview.length() > 40) {
                preview = preview.substring(0, 40) + "...";
            }
            String date = formatDate(note.updatedAt);
            
            builder.append("\n");
            builder.append("  [").append(note.id).append("] ").append(note.title).append("\n");
            builder.append("       ").append(preview).append("\n");
            builder.append("       Updated: ").append(date).append("\n");
            if (!note.tags.isEmpty()) {
                builder.append("       Tags: ").append(listToString(note.tags)).append("\n");
            }
        }

        return builder.toString();
    }

    /**
     * Shows a specific note
     */
    private String showNote(Context context, ArrayList<String> args) {
        if (args.isEmpty()) {
            return "Usage: note show <id or title>";
        }

        String query = listToString(args);
        Note note = findNote(context, query);

        if (note == null) {
            return "Note not found: " + query + "\nUse 'note list' to see all notes.";
        }

        StringBuilder builder = new StringBuilder();
        builder.append("\n");
        builder.append("═══════════════════════════════════════════════════════════════\n");
        builder.append("                    NOTE: ").append(note.title).append("                       \n");
        builder.append("═══════════════════════════════════════════════════════════════\n");
        builder.append("\n");
        builder.append("ID: ").append(note.id).append("\n");
        builder.append("Created: ").append(formatDate(note.createdAt)).append("\n");
        builder.append("Updated: ").append(formatDate(note.updatedAt)).append("\n");
        if (!note.tags.isEmpty()) {
            builder.append("Tags: ").append(listToString(note.tags)).append("\n");
        }
        builder.append("\n");
        builder.append("─".repeat(60)).append("\n");
        builder.append("\n");
        builder.append(note.content).append("\n");
        builder.append("\n");
        builder.append("─".repeat(60)).append("\n");
        builder.append("\n");
        builder.append("Use 'note edit ").append(note.id).append("' to modify or 'note delete ").append(note.id).append("' to remove");

        return builder.toString();
    }

    /**
     * Edits a note's content
     */
    private String editNote(Context context, ArrayList<String> args) {
        if (args.size() < 2) {
            return "Usage: note edit <id> <new content>";
        }

        String id = args.get(0);
        StringBuilder contentBuilder = new StringBuilder();
        for (int i = 1; i < args.size(); i++) {
            if (i > 1) contentBuilder.append(" ");
            contentBuilder.append(args.get(i));
        }
        String newContent = contentBuilder.toString();

        ArrayList<Note> notes = getAllNotes(context);
        int index = -1;
        for (int i = 0; i < notes.size(); i++) {
            Note note = notes.get(i);
            if (note.id.equals(id) || note.title.toLowerCase().equals(id.toLowerCase())) {
                index = i;
                break;
            }
        }

        if (index == -1) {
            return "Note not found: " + id;
        }

        Note oldNote = notes.get(index);
        ArrayList<String> newTags = extractTags(newContent);
        Note updatedNote = new Note(
            oldNote.id,
            oldNote.title,
            newContent,
            oldNote.createdAt,
            System.currentTimeMillis(),
            newTags
        );
        notes.set(index, updatedNote);

        saveAllNotes(context, notes);
        return "Note updated: [" + updatedNote.id + "] " + updatedNote.title;
    }

    /**
     * Deletes a note
     */
    private String deleteNote(Context context, ArrayList<String> args) {
        if (args.isEmpty()) {
            return "Usage: note delete <id>";
        }

        String id = args.get(0);
        ArrayList<Note> notes = getAllNotes(context);
        int index = -1;
        for (int i = 0; i < notes.size(); i++) {
            Note note = notes.get(i);
            if (note.id.equals(id) || note.title.toLowerCase().equals(id.toLowerCase())) {
                index = i;
                break;
            }
        }

        if (index == -1) {
            return "Note not found: " + id;
        }

        Note removed = notes.remove(index);
        saveAllNotes(context, notes);
        return "Note deleted: " + removed.title;
    }

    /**
     * Searches notes
     */
    private String searchNotes(Context context, ArrayList<String> args) {
        if (args.isEmpty()) {
            return "Usage: note search <query>";
        }

        String query = listToString(args).toLowerCase();
        ArrayList<Note> allNotes = getAllNotes(context);
        ArrayList<Note> results = new ArrayList<>();

        for (Note note : allNotes) {
            if (note.title.toLowerCase().contains(query) ||
                note.content.toLowerCase().contains(query) ||
                containsTag(note.tags, query)) {
                results.add(note);
            }
        }

        if (results.isEmpty()) {
            return "No notes found matching: " + query;
        }

        StringBuilder builder = new StringBuilder();
        builder.append("\n");
        builder.append("Search results for \"").append(query).append("\" (").append(results.size()).append(" found):\n");
        builder.append("─".repeat(60)).append("\n");

        for (Note note : results) {
            String preview = note.content.replace("\n", " ");
            if (preview.length() > 60) {
                preview = preview.substring(0, 60);
            }
            builder.append("  [").append(note.id).append("] ").append(note.title).append("\n");
            builder.append("       ").append(preview).append("\n");
        }

        return builder.toString();
    }

    private boolean containsTag(ArrayList<String> tags, String query) {
        for (String tag : tags) {
            if (tag.toLowerCase().contains(query)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Adds a tag to a note
     */
    private String tagNote(Context context, ArrayList<String> args) {
        if (args.size() < 2) {
            return "Usage: note tag <id> <tag>";
        }

        String id = args.get(0);
        String tag = args.get(1).toLowerCase();

        ArrayList<Note> notes = getAllNotes(context);
        int index = -1;
        for (int i = 0; i < notes.size(); i++) {
            Note note = notes.get(i);
            if (note.id.equals(id) || note.title.toLowerCase().equals(id.toLowerCase())) {
                index = i;
                break;
            }
        }

        if (index == -1) {
            return "Note not found: " + id;
        }

        Note note = notes.get(index);
        if (note.tags.contains(tag)) {
            return "Tag already exists: " + tag;
        }

        note.tags.add(tag);
        saveAllNotes(context, notes);
        return "Tag added: " + tag + " to note [" + note.id + "]";
    }

    /**
     * Exports a note
     */
    private String exportNote(Context context, ArrayList<String> args) {
        if (args.isEmpty()) {
            return "Usage: note export <id>";
        }

        String id = args.get(0);
        Note note = findNote(context, id);

        if (note == null) {
            return "Note not found: " + id;
        }

        StringBuilder builder = new StringBuilder();
        builder.append("Title: ").append(note.title).append("\n");
        builder.append("Created: ").append(formatDate(note.createdAt)).append("\n");
        builder.append("Updated: ").append(formatDate(note.updatedAt)).append("\n");
        if (!note.tags.isEmpty()) {
            builder.append("Tags: ").append(listToString(note.tags)).append("\n");
        }
        builder.append("\n");
        builder.append(note.content).append("\n");

        return builder.toString();
    }

    /**
     * Clears all notes
     */
    private String clearAllNotes(Context context) {
        saveAllNotes(context, new ArrayList<>());
        return "All notes cleared.";
    }

    private ArrayList<Note> getAllNotes(Context context) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        String json = prefs.getString(KEY_NOTES, "[]");
        return parseNotesJson(json != null ? json : "[]");
    }

    private Note findNote(Context context, String query) {
        ArrayList<Note> notes = getAllNotes(context);
        for (Note note : notes) {
            if (note.id.equals(query) || note.title.toLowerCase().equals(query.toLowerCase())) {
                return note;
            }
        }
        return null;
    }

    private void saveNote(Context context, Note note) {
        ArrayList<Note> notes = getAllNotes(context);
        notes.add(note);
        saveAllNotes(context, notes);
    }

    private void saveAllNotes(Context context, ArrayList<Note> notes) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        String json = notesToJson(notes);
        prefs.edit().putString(KEY_NOTES, json).apply();
    }

    private String generateNoteId() {
        String chars = "abcdefghijklmnopqrstuvwxyz0123456789";
        StringBuilder id = new StringBuilder();
        for (int i = 0; i < 6; i++) {
            int index = (int) (Math.random() * chars.length());
            id.append(chars.charAt(index));
        }
        return id.toString();
    }

    private ArrayList<String> extractTags(String content) {
        ArrayList<String> tags = new ArrayList<>();
        Pattern pattern = Pattern.compile("#(\\w+)");
        Matcher matcher = pattern.matcher(content);
        while (matcher.find()) {
            tags.add(matcher.group(1));
        }
        return tags;
    }

    private String formatDate(long timestamp) {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault());
        return sdf.format(new Date(timestamp));
    }

    private String listToString(ArrayList<String> list) {
        if (list.isEmpty()) return "";
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < list.size(); i++) {
            if (i > 0) sb.append(" ");
            sb.append(list.get(i));
        }
        return sb.toString();
    }

    private ArrayList<Note> parseNotesJson(String json) {
        if (json.equals("[]") || json.trim().isEmpty()) {
            return new ArrayList<>();
        }

        try {
            ArrayList<Note> notes = new ArrayList<>();
            String cleanJson = json.trim();
            if (cleanJson.startsWith("[")) cleanJson = cleanJson.substring(1);
            if (cleanJson.endsWith("]")) cleanJson = cleanJson.substring(0, cleanJson.length() - 1);
            
            if (cleanJson.trim().isEmpty()) return notes;

            String[] segments = cleanJson.split("\\},\\{");
            for (String segment : segments) {
                String clean = segment.trim();
                if (clean.startsWith("{")) clean = clean.substring(1);
                if (clean.endsWith("}")) clean = clean.substring(0, clean.length() - 1);
                
                String[] parts = clean.split(",(?=\")");
                
                String id = "";
                String title = "";
                String content = "";
                long created = 0L;
                long updated = 0L;
                ArrayList<String> tags = new ArrayList<>();

                for (String part : parts) {
                    String trimmed = part.trim();
                    if (trimmed.startsWith("\"id\":")) {
                        id = trimmed.substring(4).trim().replace("\"", "");
                    } else if (trimmed.startsWith("\"title\":")) {
                        title = trimmed.substring(7).trim().replace("\"", "");
                    } else if (trimmed.startsWith("\"content\":")) {
                        content = trimmed.substring(9).trim().replace("\"", "");
                    } else if (trimmed.startsWith("\"created\":")) {
                        try {
                            created = Long.parseLong(trimmed.substring(9).trim());
                        } catch (NumberFormatException e) {
                            // Use default
                        }
                    } else if (trimmed.startsWith("\"updated\":")) {
                        try {
                            updated = Long.parseLong(trimmed.substring(9).trim());
                        } catch (NumberFormatException e) {
                            // Use default
                        }
                    } else if (trimmed.startsWith("\"tags\":")) {
                        String tagsStr = trimmed.substring(7).trim();
                        if (tagsStr.startsWith("[")) tagsStr = tagsStr.substring(1);
                        if (tagsStr.endsWith("]")) tagsStr = tagsStr.substring(0, tagsStr.length() - 1);
                        if (!tagsStr.trim().isEmpty()) {
                            String[] tagArray = tagsStr.split(",");
                            for (String tag : tagArray) {
                                tags.add(tag.trim().replace("\"", ""));
                            }
                        }
                    }
                }

                notes.add(new Note(id, title, content, created, updated, tags));
            }

            return notes;
        } catch (Exception e) {
            Log.e(TAG, "Error parsing notes: " + e.getMessage(), e);
            return new ArrayList<>();
        }
    }

    private String notesToJson(ArrayList<Note> notes) {
        StringBuilder json = new StringBuilder("[");
        for (int i = 0; i < notes.size(); i++) {
            if (i > 0) json.append(",");
            Note note = notes.get(i);
            json.append("{");
            json.append("\"id\":\"").append(note.id).append("\",");
            json.append("\"title\":\"").append(escape(note.title)).append("\",");
            json.append("\"content\":\"").append(escape(note.content)).append("\",");
            json.append("\"created\":").append(note.createdAt).append(",");
            json.append("\"updated\":").append(note.updatedAt).append(",");
            json.append("\"tags\":[");
            for (int j = 0; j < note.tags.size(); j++) {
                if (j > 0) json.append(",");
                json.append("\"").append(note.tags.get(j)).append("\"");
            }
            json.append("]");
            json.append("}");
        }
        json.append("]");
        return json.toString();
    }

    private String escape(String input) {
        if (input == null) return "";
        return input.replace("\\", "\\\\")
                    .replace("\"", "\\\"")
                    .replace("\n", "\\n")
                    .replace("\r", "\\r")
                    .replace("\t", "\\t");
    }
}

File: tui-smart-launcher/tui/smartlauncher/productivity/NotesCommand.kt
package tui.smartlauncher.productivity

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Notes Command - Quick note taking and retrieval
 * Supports creating, listing, searching, and managing notes
 */
class NotesCommand : CommandHandler {

    companion object {
        private const val TAG = "NotesCommand"
        private const val PREFS_NAME = "tui_notes_prefs"
        private const val KEY_NOTES = "saved_notes"
        private const val MAX_NOTE_LENGTH = 10000
    }

    data class Note(
        val id: String,
        val title: String,
        val content: String,
        val createdAt: Long,
        val updatedAt: Long,
        val tags: List<String>
    )

    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    override fun getName(): String = "note"

    override fun getAliases(): List<String> = listOf("notes", "notepad", "memo", "jot")

    override fun getDescription(): String = "Quick note taking and management"

    override fun getUsage(): String = """
        ╔══════════════════════════════════════════════════════╗
        ║                   NOTES COMMANDS                      ║
        ╠══════════════════════════════════════════════════════╣
        ║  note create <title>         - Create new note       ║
        ║  note create <title> <text>  - Create with content   ║
        ║  note list                   - List all notes        ║
        ║  note show <id>              - Show note content     ║
        ║  note show <title>           - Show by title         ║
        ║  note edit <id> <text>       - Edit note content     ║
        ║  note delete <id>            - Delete a note         ║
        ║  note search <query>         - Search notes          ║
        ║  note tag <id> <tag>         - Add tag to note       ║
        ║  note export <id>            - Export note as text   ║
        ╚══════════════════════════════════════════════════════╝
    """.trimIndent()

    override fun execute(context: Context, args: List<String>): String {
        if (args.isEmpty()) {
            return listNotes()
        }

        val command = args[0].lowercase()
        val parameters = args.drop(1)

        return when (command) {
            "create", "new", "add" -> createNote(parameters)
            "list", "ls", "ls" -> listNotes()
            "show", "view", "get" -> showNote(parameters)
            "edit", "update", "modify" -> editNote(parameters)
            "delete", "del", "remove" -> deleteNote(parameters)
            "search", "find", "grep" -> searchNotes(parameters)
            "tag", "tags" -> tagNote(parameters)
            "export" -> exportNote(parameters)
            "clear" -> clearAllNotes()
            else -> {
                // Try to find note by title or ID
                showNote(listOf(command))
            }
        }
    }

    /**
     * Creates a new note
     */
    private fun createNote(args: List<String>): String {
        if (args.isEmpty()) {
            return "Usage: note create <title> [content]\nUse 'note create \"My Title\" This is content' for inline creation."
        }

        val title = args[0]
        val content = if (args.size > 1) args.drop(1).joinToString(" ") else ""

        return try {
            val note = Note(
                id = generateNoteId(),
                title = title,
                content = content,
                createdAt = System.currentTimeMillis(),
                updatedAt = System.currentTimeMillis(),
                tags = extractTags(content)
            )

            saveNote(note)
            "Note created: [${note.id}] ${note.title}\n${if (content.isNotBlank()) "Content: ${content.take(50)}${if (content.length > 50) "..." else ""}" else "(Empty note - use 'note edit ${note.id}' to add content)"}"
        } catch (e: Exception) {
            "Error creating note: ${e.message}"
        }
    }

    /**
     * Lists all notes
     */
    private fun listNotes(): String {
        val notes = getAllNotes()
        if (notes.isEmpty()) {
            return buildString {
                appendLine()
                appendLine("No notes yet.")
                appendLine()
                appendLine("Create your first note:")
                appendLine("  note create \"My Title\" This is the content")
            }
        }

        val builder = StringBuilder()
        builder.appendLine()
        builder.appendLine("═══════════════════════════════════════════════════════════════")
        builder.appendLine("                         NOTES (${notes.size})                            ")
        builder.appendLine("═══════════════════════════════════════════════════════════════")

        notes.sortedByDescending { it.updatedAt }.forEachIndexed { index, note ->
            val preview = note.content.take(40).replace("\n", " ")
            val date = formatDate(note.updatedAt)
            builder.appendLine()
            builder.appendLine("  [${note.id}] ${note.title}")
            builder.appendLine("       $preview${if (note.content.length > 40) "..." else ""}")
            builder.appendLine("       Updated: $date")
            if (note.tags.isNotEmpty()) {
                builder.appendLine("       Tags: ${note.tags.joinToString(", ")}")
            }
        }

        return builder.toString()
    }

    /**
     * Shows a specific note
     */
    private fun showNote(args: List<String>): String {
        if (args.isEmpty()) {
            return "Usage: note show <id or title>"
        }

        val query = args.joinToString(" ")
        val note = findNote(query)

        if (note == null) {
            return "Note not found: $query\nUse 'note list' to see all notes."
        }

        return buildString {
            appendLine()
            appendLine("═══════════════════════════════════════════════════════════════")
            appendLine("                    NOTE: ${note.title}                       ")
            appendLine("═══════════════════════════════════════════════════════════════")
            appendLine()
            appendLine("ID: ${note.id}")
            appendLine("Created: ${formatDate(note.createdAt)}")
            appendLine("Updated: ${formatDate(note.updatedAt)}")
            if (note.tags.isNotEmpty()) {
                appendLine("Tags: ${note.tags.joinToString(", ")}")
            }
            appendLine()
            appendLine("─".repeat(60))
            appendLine()
            appendLine(note.content)
            appendLine()
            appendLine("─".repeat(60))
            appendLine()
            appendLine("Use 'note edit ${note.id}' to modify or 'note delete ${note.id}' to remove")
        }
    }

    /**
     * Edits a note's content
     */
    private fun editNote(args: List<String>): String {
        if (args.size < 2) {
            return "Usage: note edit <id> <new content>"
        }

        val id = args[0]
        val newContent = args.drop(1).joinToString(" ")

        val notes = getAllNotes().toMutableList()
        val index = notes.indexOfFirst { it.id == id || it.title.lowercase() == id.lowercase() }

        if (index == -1) {
            return "Note not found: $id"
        }

        val oldNote = notes[index]
        val updatedNote = oldNote.copy(
            content = newContent,
            updatedAt = System.currentTimeMillis(),
            tags = extractTags(newContent)
        )
        notes[index] = updatedNote

        saveAllNotes(notes)
        return "Note updated: [${updatedNote.id}] ${updatedNote.title}"
    }

    /**
     * Deletes a note
     */
    private fun deleteNote(args: List<String>): String {
        if (args.isEmpty()) {
            return "Usage: note delete <id>"
        }

        val id = args[0]
        val notes = getAllNotes().toMutableList()
        val index = notes.indexOfFirst { it.id == id || it.title.lowercase() == id.lowercase() }

        if (index == -1) {
            return "Note not found: $id"
        }

        val removed = notes.removeAt(index)
        saveAllNotes(notes)
        return "Note deleted: ${removed.title}"
    }

    /**
     * Searches notes
     */
    private fun searchNotes(args: List<String>): String {
        if (args.isEmpty()) {
            return "Usage: note search <query>"
        }

        val query = args.joinToString(" ").lowercase()
        val results = getAllNotes().filter {
            it.title.lowercase().contains(query) ||
            it.content.lowercase().contains(query) ||
            it.tags.any { tag -> tag.lowercase().contains(query) }
        }

        if (results.isEmpty()) {
            return "No notes found matching: $query"
        }

        val builder = StringBuilder()
        builder.appendLine()
        builder.appendLine("Search results for \"$query\" (${results.size} found):")
        builder.appendLine("─".repeat(60))

        results.forEach { note ->
            builder.appendLine("  [${note.id}] ${note.title}")
            builder.appendLine("       ${note.content.take(60).replace("\n", " ")}")
        }

        return builder.toString()
    }

    /**
     * Adds a tag to a note
     */
    private fun tagNote(args: List<String>): String {
        if (args.size < 2) {
            return "Usage: note tag <id> <tag>"
        }

        val id = args[0]
        val tag = args[1].lowercase()

        val notes = getAllNotes().toMutableList()
        val index = notes.indexOfFirst { it.id == id || it.title.lowercase() == id.lowercase() }

        if (index == -1) {
            return "Note not found: $id"
        }

        val note = notes[index]
        if (note.tags.contains(tag)) {
            return "Tag already exists: $tag"
        }

        notes[index] = note.copy(tags = note.tags + tag)
        saveAllNotes(notes)
        return "Tag added: $tag to note [${note.id}]"
    }

    /**
     * Exports a note
     */
    private fun exportNote(args: List<String>): String {
        if (args.isEmpty()) {
            return "Usage: note export <id>"
        }

        val id = args[0]
        val note = findNote(id)

        if (note == null) {
            return "Note not found: $id"
        }

        return buildString {
            appendLine("Title: ${note.title}")
            appendLine("Created: ${formatDate(note.createdAt)}")
            appendLine("Updated: ${formatDate(note.updatedAt)}")
            if (note.tags.isNotEmpty()) {
                appendLine("Tags: ${note.tags.joinToString(", ")}")
            }
            appendLine()
            appendLine(note.content)
        }
    }

    /**
     * Clears all notes
     */
    private fun clearAllNotes(): String {
        saveAllNotes(emptyList())
        return "All notes cleared."
    }

    private fun getAllNotes(): List<Note> {
        val json = prefs.getString(KEY_NOTES, "[]") ?: "[]"
        return parseNotesJson(json)
    }

    private fun findNote(query: String): Note? {
        return getAllNotes().find {
            it.id == query || it.title.lowercase() == query.lowercase()
        }
    }

    private fun saveNote(note: Note) {
        val notes = getAllNotes().toMutableList()
        notes.add(note)
        saveAllNotes(notes)
    }

    private fun saveAllNotes(notes: List<Note>) {
        val json = notesToJson(notes)
        prefs.edit().putString(KEY_NOTES, json).apply()
    }

    private fun generateNoteId(): String {
        val chars = "abcdefghijklmnopqrstuvwxyz0123456789"
        return (1..6).map { chars.random() }.joinToString("")
    }

    private fun extractTags(content: String): List<String> {
        val tagPattern = Regex("#(\\w+)")
        return tagPattern.findAll(content).map { it.groupValues[1] }.toList()
    }

    private fun formatDate(timestamp: Long): String {
        val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault())
        return sdf.format(Date(timestamp))
    }

    private fun parseNotesJson(json: String): List<Note> {
        if (json == "[]" || json.isBlank()) return emptyList()

        return try {
            val cleanJson = json.trim().removeSurrounding("[", "]")
            if (cleanJson.isBlank()) return emptyList()

            cleanJson.split("},{").map { segment ->
                val clean = segment.removeSurrounding("{", "}")
                val parts = clean.split(",(?=\")".toRegex())

                val id = parts.find { it.contains("id") }?.substringAfter(":")?.trim()?.removeSurrounding("\"") ?: ""
                val title = parts.find { it.contains("title") }?.substringAfter(":")?.trim()?.removeSurrounding("\"") ?: ""
                val content = parts.find { it.contains("content") }?.substringAfter(":")?.trim()?.removeSurrounding("\"") ?: ""
                val created = parts.find { it.contains("created") }?.substringAfter(":")?.trim()?.toLongOrNull() ?: 0L
                val updated = parts.find { it.contains("updated") }?.substringAfter(":")?.trim()?.toLongOrNull() ?: 0L

                Note(id, title, content, created, updated, emptyList())
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error parsing notes: ${e.message}")
            emptyList()
        }
    }

    private fun notesToJson(notes: List<Note>): String {
        return notes.joinToString(",", "[", "]") { note ->
            buildString {
                append("{")
                append("\"id\":\"${note.id}\",")
                append("\"title\":\"${escape(note.title)}\",")
                append("\"content\":\"${escape(note.content)}\",")
                append("\"created\":${note.createdAt},")
                append("\"updated\":${note.updatedAt},")
                append("\"tags\":[${note.tags.map { "\"$it\"" }.joinToString(",")}]")
                append("}")
            }
        }
    }

    private fun escape(input: String): String {
        return input.replace("\\", "\\\\")
            .replace("\"", "\\\"")
            .replace("\n", "\\n")
            .replace("\r", "\\r")
            .replace("\t", "\\t")
    }
}
